### PR TITLE
add shogun recipe [ready despite the X]

### PR DESCRIPTION
**Pull request by dougalsutherland**
Adds a recipe for the shogun base library, and a python-shogun recipe for the python interface. (Could potentially make a recipe for the R interface too.)

Non-blocking problems that I'll turn into issues in the feedstock repo, once it exists:
- Windows build issues (so this version skips it):
  - Shogun won't build with VC < 14 or on win32, so python versions < 3.5 are impossible anyway.
  - Shogun's cmake setup doesn't handle optional shared library dependencies correctly yet: https://github.com/shogun-toolbox/shogun/issues/3917
  - Even if I turn off all the optional shared lib dependencies on Windows, AppVeyor is running out of disk space ([example build](https://ci.appveyor.com/project/conda-forge/staged-recipes/build/1.0.14895/job/j7175cu857satqxa)).
  - BLAS/LAPACK support (optional but helpful) would need to build with MKL; is that possible on conda-forge?
- OpenMP support on Mac: requires new clang (from homebrew) or gcc
   - gcc in `defaults` is too old: I get [this bug](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=66509), fixed in gcc 5.2 per https://github.com/OpenAADL/ocarina/issues/60#issuecomment-174658708, but `defaults` has 4.8.5
   - So, seems impossible unless conda-forge wants to start packaging new gcc / clang. (Would also require `toolchain` changes.)
- Tell Eigen to use OpenBLAS: will be supported in a future version of Shogun, https://github.com/shogun-toolbox/shogun/issues/3914

### CODE REVIEW

**Comment by dougalsutherland** [comment]
Okay, I'm getting timeouts on Linux/Mac (not entirely unexpected, the library takes a long time to build...) and build errors on Windows.

I'll try to look into building a base shogun package and then a shogun-python package for the interface, so that the library is compiled once per arch instead of 6 times.

For Windows:
- the VC 9 ones (python 2.7) are complaining about something about the build system: 
```
(C:\bld\shogun_1499393886733\_b_env) C:\bld\shogun_1499393886733\work\shogun-shogun_6.0.0\build>msbuild src\shogun\libshogun.vcxproj 
Microsoft (R) Build Engine Version 2.0.50727.8007
[Microsoft .NET Framework, Version 2.0.50727.8669]
Copyright (C) Microsoft Corporation 2005. All rights reserved.
MSBUILD : error MSB1009: Project file does not exist.
Switch: src\shogun\libshogun.vcxproj
```

- the higher VCs are giving errors like
```
C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\limits(210): warning C4577: 'noexcept' used with no exception handling mode specified; termination on exception is not guaranteed. Specify /EHsc [C:\bld\shogun_1499394521637\work\shogun-shogun_6.0.0\build\src\shogun\libshogun.vcxproj]
```
so I guess some compile flags aren't being set appropriately.

**Comment by dougalsutherland** [comment]
Why is appveyor not being run? I have a:
```
skip: True  # [win and py2k]
```
line because shogun won't build on VC9, but there's a `python  # [win]` build requirement...

**Comment by karlnapf** [comment]
@vigsterkr

**Comment by vigsterkr** [comment]
@karlnapf yep, we've been in touch with @dougalsutherland during the weekend... there are couple of things to be fixed on windows build...

**Comment by dougalsutherland** [comment]
Hmm...after turning off all the optional dependencies on Windows to avoid https://github.com/shogun-toolbox/shogun/issues/3917, the AppVeyor build is now running out of disk space, lol. Not sure what to do about that....

**Comment by jschueller** [comment]
- on windows CMAKE_PREFIX_PATH should be set to LIBRARY_PREFIX
- to successfully link to openblas I had to add prefix/lib/openblas.so to the list of linked libs while avoidind to call the findlapack/findbals cmake macros

**Comment by dougalsutherland** [comment]
@jschueller Was this after including `use-detected-lapack.patch`? I noticed that this patch makes it work on OS X but it links to both `libopenblas` and also `libblas`/`liblapack` on Linux; haven't had a chance to track that down yet.

**Comment by jschueller** [comment]
no idea

**Comment by jschueller** [comment]
maybe some dependency links to it?

**Comment by dougalsutherland** [comment]
The built package wasn't actually linking to libblas/liblapack; `ldd` just thought it was when I was looking at the `.so` without it being installed in a conda env, I guess because some other dependency was resolving to a system version that used that. So the Linux/Mac portions of this recipe are ready, I think.

**Comment by dougalsutherland** [comment]
@vigsterkr and @karlnapf: I think this is ready for Linux/Mac, so if I'm messing something up or violating a license or anything, now's the time to say so. :)

**Comment by jschueller** [comment]
The users expects "conda install shogun" to provide the bindings. I'd rather have a package named shogun with the bindings, and a shogun-cpp package for the c library.

**Comment by jschueller** [comment]
If you can afford to have a single package for both (ie enough time constraints on cis), imho it's not worth splitting it because harder to maintain.

**Comment by dougalsutherland** [comment]
@jschueller Sure, I just did that following e.g. leveldb / python-leveldb, blosc / python-blosc, etc. How about libshogun instead of shogun-cpp?

**Comment by jschueller** [comment]
I proposed shogun-cpp as we have boost-cpp, the idea is the same.

**Comment by dougalsutherland** [comment]
I split it because building libshogun 6 times was timing out on CircleCI / Travis. Shogun likes ccache for these things, but I don't think that plays nicely with `conda-build-all` since it's using a different `ccache` package for each; there might be some way to hack it....

**Comment by jschueller** [comment]
Do you know cotire ?

**Comment by dougalsutherland** [comment]
@jschueller Haven't seen it before. Does it work in separate conda envs / is there an existing conda-forge recipe that uses it?

**Comment by jschueller** [comment]
The build system must be patched to use cotire, openturns uses it.

I see there's a ccache-toolchain package to use ccache.

**Comment by dougalsutherland** [comment]
ccache-toolchain, running locally, is giving me very few cache hits (< 3% after running two builds). Something about the different environment names, maybe.

A quick attempt at adding cotire to shogun is giving me lots of compile errors.

I'll change the recipe names....

**Comment by isuruf** [comment]
`ccache-toolchain` is incomplete without https://github.com/conda-forge/conda-smithy/pull/526.

**Comment by dougalsutherland** [comment]
@isuruf, setting `CCACHE_BASEDIR` in the recipe (mentioned in that PR) got me up to 30% for running `conda-build-all` locally; not sure if that's enough for builds to finish. Do you know if the PR would help any more than that?

**Comment by dougalsutherland** [comment]
Travis timed out on the very last distribution. :(

This should be fine once it's rendered, though, because `shogun-cpp` will be in a separate feedstock before the `shogun` recipe, and both jobs should have enough time then. So I think this is ready.

**Comment by dougalsutherland** [comment]
@jschueller, sorry to bug but any thoughts? The shogun people are eager to get this merged. :)

File: recipes/shogun-cpp/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    - glpk
+    - hdf5
+    - json-c
+    - libxml2
+    - lzo
+    - openblas
+    - protobuf
+    - snappy
+    - xz
+    - zlib
```
Comment (isuruf) [inline recipes/shogun-cpp/meta.yaml]: You need to pin the dependencies above. See https://github.com/conda-forge/conda-forge.github.io/blob/master/scripts/pin_the_slow_way.py for some of them.

File: recipes/shogun/build.sh [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    -DENABLE_TESTING=OFF \
+    -DENABLE_COVERAGE=OFF \
+    -DUSE_SVMLIGHT=OFF \
+    -DSWIG_EXECUTABLE=$PREFIX/bin/swig \
+    -DLIBSHOGUN=OFF \
+    -DPYTHON_INCLUDE_DIR=$pyinc \
+    -DPYTHON_LIBRARY=$pylib \
+    -DPYTHON_EXECUTABLE=$PYTHON \
+    -DPythonModular=ON
+
```
Comment (isuruf) [inline recipes/shogun/build.sh]: Add `make -j${CPU_COUNT}` here.

Comment (isuruf) [inline recipes/shogun/build.sh]: Can you run the tests here? (Doesn't have to be the full test-suite if it's long.)

Comment (dougalsutherland) [inline recipes/shogun/build.sh]: Won't change anything, but sure.

File: recipes/shogun-cpp/meta.yaml (hunk 1 above)
Comment (dougalsutherland) [inline recipes/shogun-cpp/meta.yaml]: So should every dependency be pinned? To what, the current most recent version if it's not there?

Comment (isuruf) [inline recipes/shogun-cpp/meta.yaml]: Every C/C++ library should be pinned. If it's not there, `major.minor` of the most recent version is a good bet, but that depends on the package. There are some reports here https://abi-laboratory.pro/tracker/ about ABI compatibility. What are the dependencies not there?

Comment (dougalsutherland) [inline recipes/shogun-cpp/meta.yaml]: Packages not in `pin_the_slow_way` that get linked against:
- arpack
- curl
- glpk
- json-c
- lzo
- protobuf
- snappy

Comment (isuruf) [inline recipes/shogun-cpp/meta.yaml]: Okay, `major.minor` is okay I guess. For snappy, full version is needed.

Comment (dougalsutherland) [inline recipes/shogun-cpp/meta.yaml]: Okay, I took reasonable guesses at the needed versions for those (based on what's in conda-forge + abi-laboratory.pro) and pinned them.

Comment (dougalsutherland) [inline recipes/shogun-cpp/meta.yaml]: Hmm, github apparently doesn't reload these comments automatically so I didn't see what you were saying @isuruf. I came to basically the same conclusions though.

**Comment by dougalsutherland** [comment]
@isuruf Re: tests (again not seeing those comments until a full page reload...), the full test suite is huge and takes a while to build. I have some sanity checks in `run_test.py`; is there any advantage to checking in the build phase rather than the test phase?

File: recipes/shogun/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    - shogun-cpp {{ version }}
+    - swig
+  run:
+    - python
+    - numpy x.x
+    - shogun-cpp {{ version }}
+
+test:
+  requirements:
+    - numpy x.x
```
Comment (isuruf) [inline recipes/shogun/meta.yaml]: You can remove the above 2 lines

**Comment by isuruf** [comment]
> I have some sanity checks in run_test.py; is there any advantage to checking in the build phase rather than the test phase?

I didn't see that there was a `run_test.py` for `shogun-cpp` as well, then that's fine.

File: recipes/shogun-cpp/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    - curl 7.52.*
+    - eigen
+    - glpk 4.61|4.62
+    - hdf5 1.8.18|1.8.18.*
+    - json-c 0.12|0.12.*
+    - libxml2 2.9.*
+    - lzo 2.09
+    - openblas 0.2.19|0.2.19.*
+    - protobuf 3.3.*
+    - snappy 1.1.4|1.1.6
```
Comment (isuruf) [inline recipes/shogun-cpp/meta.yaml]: There are new symbols in `1.1.6` than `1.1.4`. So it's better to use `1.1.6` here.

Comment (dougalsutherland) [inline recipes/shogun-cpp/meta.yaml]: 1.1.6 isn't on conda-forge yet, though. According to the release notes, 1.1.6 is basically the same as 1.1.4 just with some build system changes; 1.1.5 was broken in the meantime.

Comment (isuruf) [inline recipes/shogun-cpp/meta.yaml]: yes, 1.1.6 is 1.1.5 with soversion fixed, but there are some new symbols and removed symbols in 1.1.5 wrt 1.1.4 and hence these changes are there in 1.1.6 wrt 1.1.4. https://abi-laboratory.pro/tracker/compat_report/snappy/1.1.4/1.1.5/48504/abi_compat_report.html#Added

So, just 1.1.4 is better here.

Comment (dougalsutherland) [inline recipes/shogun-cpp/meta.yaml]: Ah, I see, I trusted the changelog without looking more closely. :)  Changed to just 1.1.4.

**Comment by karlnapf** [comment]
For tests, you could exclude unit tests, libshogun example tests, and integration  tests, leaving only executing the meta examples (in c++ and python)

Check the switches that start with TRAVIS_ for that

**Comment by karlnapf** [comment]
Or even only run the python meta examples, to drop the need to compile something

**Comment by dougalsutherland** [comment]
@karlnapf Building meta-examples requires ply and stuff, which I'm skipping here. Right now I have I'm using `BUILD_EXAMPLES=ON` and [running a handful of them](https://github.com/dougalsutherland/staged-recipes/blob/c13a2f8998454beff39312496d0f24f9757d3454/recipes/shogun-cpp/run_test.py). Is that reasonable?

**Comment by dougalsutherland** [comment]
FYI for others, @karlnapf and I talked offline and decided that the examples are good for now, but a future shogun release will include some more-minimal minor check code that serves that purpose.

**Comment by dougalsutherland** [comment]
Thanks @isuruf!

**Comment by isuruf** [comment]
You'll have to rerender or restart the CI of `shogun-feedstock` because `shogun-cpp` will be missing when `shogun-feedstock` is created.

**Pull request merged**